src/OAuth/OAuth1/Service/Geocaching.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+namespace OAuth\OAuth1\Service;
+
+use OAuth\OAuth1\Signature\SignatureInterface;
+use OAuth\OAuth1\Token\StdOAuth1Token;
+use OAuth\Common\Http\Exception\TokenResponseException;
+use OAuth\Common\Http\Uri\Uri;
+use OAuth\Common\Consumer\CredentialsInterface;
+use OAuth\Common\Http\Uri\UriInterface;
+use OAuth\Common\Storage\TokenStorageInterface;
+use OAuth\Common\Http\Client\ClientInterface;
+use OAuth\Common\Exception\Exception;
+
+class Geocaching extends AbstractService
+{
+    public function __construct(
+        CredentialsInterface $credentials,
+        ClientInterface $httpClient,
+        TokenStorageInterface $storage,
+        SignatureInterface $signature,
+        UriInterface $baseApiUri = null
+    ) {
+        parent::__construct($credentials, $httpClient, $storage, $signature, $baseApiUri);
+        if ($baseApiUri === null) {
+            $this->setBaseApiUri();
+        }
+        $this->setEndPoint();
+    }
+
+    public function getRequestTokenEndpoint()
+    {
+        return $this->getEndPoint();
+    }
+
+    public function getAuthorizationEndpoint()
+    {
+        return $this->getEndPoint();
+    }
+
+    public function getAccessTokenEndpoint()
+    {
+        return $this->getEndPoint();
+    }
+
+    public function getEndPoint()
+    {
+        return new Uri($this->endPoint);
+    }
+
+    public function setEndPoint($endPoint = 'live')
+    {
+        switch($endPoint) {
+            case 'staging':
```

Comment:
Sorry, the exact line I commented on was confusing. When somebody uses this class the possible values for `$endPoint` are unknown unless you look into this class and understand what it's doing - this is unnecessary logic because they can provide the URL directly:

```php
public function setEndPoint($endPoint = self::ENDPOINT_LIVE)
{
    $this->endPoint = $endPoint;
}
```

The explicit advantages are:

1, This is clear to developers as they can now expect that `Geocaching::ENDPOINT_*` are all valid values without needing to inspect the class.
2. Based on common conventions developers would also expect that a setter works as a setter. If they were to invoke `setEndPoint('https://www.geocaching.com/oauth/other/endpoint');` they would be shocked to see thats it's still pointed at `live` - this is a potential bug.
3. New endpoints do not require code logic to change.
4. Less code and less complexity = easier to maintain.